### ISSUE TITLE
Import QMServices Manually In Swift Project

### ISSUE THREAD

**Issue by yaodong-li** [QuickBlox/q-municate-services-ios#177]
Hi,
Thanks for this amazing library. I have an existing swift project and I would like to import QMServices manually because I have to use "use_framework" for my cocoapods. I followed the tutorials and faced liner error. The description of your tutorial mentioned that "Don't forget to add Quickblox and Bolts frameworks to your project.". But I'm facing linker error  which told me that "Bolts" framework can't be found. After I removed "bolts" framework form "link binary with libraries", everything is fine again. Could you tell me whether I was doing the correct thing? Please see my attached 4 mins video for details. Thanks a lot!
https://www.youtube.com/watch?v=pNPKBPk27XU

**Comment by yaodong-li** [comment]
Unfortunately, problems happens after I use QBSettings in app delegates:

<img width="349" alt="screenshot 2016-04-09 18 26 38" src="https://cloud.githubusercontent.com/assets/6023208/14403201/c8a267e4-fe80-11e5-82e4-ec4c2d70b2ce.png">

**Comment by yaodong-li** [comment]
After several hours of attempting and adding the frameworks below, the project can be built successfully.
CFNetwork.framework
libxml2.tbd
libresolv.tbd
MobileCoreServices.framework
SystemConfiguration.framework

Until I start to use functions related to BFTasks....
<img width="558" alt="screenshot 2016-04-09 20 44 02" src="https://cloud.githubusercontent.com/assets/6023208/14403680/d50b931c-fe93-11e5-97e9-99e6136686a7.png">

Is there any tips on installing qmservices on swift projects?

**Comment by Raikerian** [comment]
> After I removed "bolts" framework form "link binary with libraries", everything is fine again.

So, is that solution working for you? Maybe you have already installed Bolts framework.

**Comment by yaodong-li** [comment]
Nope. As shown in my small video, I didn't install bolts.

**Comment by Raikerian** [comment]
You have errors for our framework, not for QMServices (at least on screenshots you have provided). Follow a detailed guide about swift installation [here](http://quickblox.com/developers/IOS-how-to-connect-Quickblox-framework#Additional_steps_for_Swift_using_Manual_Installation).

**Comment by yaodong-li** [comment]
I only followed the steps of manual installation of QMServices. I didn't install quickblox or bolts directly. I can config QBSettings in appdelegates with no error. But issue comes when I start to use the apis related to bolts. I do followed steps of qmservice's manual installation instructions.

**Comment by yaodong-li** [comment]
At the same time. I also tried another solution which is to install quickblox as dynamic framework using cocopods. Everything is fine for now. It's just not as good as QMServices because I have to implement the caching manually.

**Comment by Raikerian** [comment]
Have you tried to install Bolts framework using cocoapods? The bolts errors you are having are basically saying that there is no Bolts framework :)

**Comment by yaodong-li** [comment]
No. Quickblox and bolts are all in the "Frameworks" folder. If I've referenced it directly, I should be able to use that. Did I do anything wrong in the video during setup?

**Comment by Raikerian** [comment]
Your video is restricted from view.
Both Bolts and Quickblox frameworks are only header files without binary in that folder. Try to install Bolts using cocoapods or manually from their repository.

**Comment by yaodong-li** [comment]
Oh sorry. I made the video unlisted just now. 

I think I know what the mistake I have made is. I didn't install any binary file of quickblox and bolts manually or from cocoapods. I didn't know that they were only header files in the qmservice project. Thanks a lot for your help!

**Comment by Raikerian** [comment]
Try to do it and let us know if it works. I will update our manual guide installation if it is.

**Comment by yaodong-li** [comment]
Ok I will test it

**Comment by yaodong-li** [comment]
I just tried to install quickblox and bolts using cocoapods instead. When I start to user QMServiceManager.instance() singleton, app crashes. Here is the details.

<img width="989" alt="screenshot 2016-04-11 16 05 30" src="https://cloud.githubusercontent.com/assets/6023208/14420562/64018286-ffff-11e5-8e5c-e11dedbf1a2d.png">
<img width="977" alt="screenshot 2016-04-11 16 05 43" src="https://cloud.githubusercontent.com/assets/6023208/14420561/64002314-ffff-11e5-85be-a3bb3d5ea9d1.png">
<img width="991" alt="screenshot 2016-04-11 16 05 51" src="https://cloud.githubusercontent.com/assets/6023208/14420563/6402c61e-ffff-11e5-9058-43238762e4b5.png">
<img width="744" alt="screenshot 2016-04-11 16 06 15" src="https://cloud.githubusercontent.com/assets/6023208/14420564/64049cc8-ffff-11e5-87e0-80c1b8547fb6.png">

**Comment by yaodong-li** [comment]
I installed qmservice with manual installation. Quickblox and bolts with cocoapods. After that, I followed cocoapods instructions and add $(inherited) in both header search path and framework search path. Using QBrequest to get users with tags works fine

**Comment by Raikerian** [comment]
Can you provide me your test project?

**Comment by yaodong-li** [comment]
Ok. Here is the link: https://www.dropbox.com/s/3ltdgramq2ajk6g/TestQMAgain.zip?dl=0

**Comment by Raikerian** [comment]
@fans3210 I will investigate it in a near time and post update soon

**Comment by yaodong-li** [comment]
Thanks a lot!

**Comment by Raikerian** [comment]
@fans3210 

I'm currently working on complete manual installation fix. In order for you not to wait for me, there is a possible way to install QMServices without Quickblox framework as a dependency.

1) Place a QMServices.podspec file in your project folder with this content:

```
#
#  Be sure to run `pod spec lint QMServices.podspec --verbose --use-libraries --allow-warnings' to ensure this is a
#  valid spec and to remove all comments including this before submitting the spec.
#
#  To learn more about Podspec attributes see http://docs.cocoapods.org/specification.html
#  To see working Podspecs in the CocoaPods repo see https://github.com/CocoaPods/Specs/
#   
#  To submit use `pod trunk push QMServices.podspec --verbose --use-libraries --allow-warnings`
#


Pod::Spec.new do |s|
  s.name         = "QMServices"
  s.version      = "0.3.8"
  s.summary      = "Easy-to-use services for Quickblox SDK, for speeding up development of iOS chat applications."
  s.homepage     = "https://github.com/QuickBlox/q-municate-services-ios"
  s.license      = { :type => 'BSD', :file => '../LICENSE.txt' }
  s.authors      = {"Anton Sokolchenko" => "anton.sokolchenko@quickblox.com", "Andrey Ivanov" => "andrey.ivanov@quickblox.com", "Vitaliy Gorbachov" => "vitaliy.gorbachov@quickblox.com"}
  s.platform     = :ios, "7.0"
  s.source       = { :git => "https://github.com/QuickBlox/q-municate-services-ios.git", :tag => "#{s.version}"}
  s.source_files  = "Classes", "**/*.{h,m}"
  s.exclude_files = "Pods"
  s.frameworks = "CoreData"
  s.requires_arc = true
  s.xcconfig = { "FRAMEWORK_SEARCH_PATHS" => "$(PODS_ROOT)/QuickBlox/" }
  s.prefix_header_contents = 
  '#import <Quickblox/Quickblox.h>
#import <CoreData/CoreData.h>
#import <Quickblox/QBMulticastDelegate.h>
#import <Bolts/Bolts.h>'
  s.resource_bundle = {'QMChatCacheModel' => 'QMChatCache/QMChatCache/CoreData/QMChatServiceModel.xcdatamodeld', 'QMContactListCacheModel' => 'QMContactListCache/QMContactListCache/CoreData/QMContactListModel.xcdatamodeld', 'QMUsersCacheModel' => 'QMUsersCache/QMUsersCache/CoreData/QMUsersModel.xcdatamodeld'}
end
```

2) Modify your Podfile, and make QMServices to look into this podspec. Here how it should look now:

```
pod 'QMServices', :podspec => 'QMServices.podspec'
pod 'QuickBlox'
pod 'Bolts'
```

3) Perform `pod update` or (if doesn't work) remove Podfile.lock file, Pods folder and .xcworkspace file and perform a complete `pod install`.

Let me know if it works for you.

**Comment by yaodong-li** [comment]
Apologies again. I'm having another issue this time in a new set-up project: 
<img width="604" alt="screenshot 2016-04-13 11 19 11" src="https://cloud.githubusercontent.com/assets/6023208/14481874/94942456-0169-11e6-9258-21eb7f984243.png">

Here is my project: https://www.dropbox.com/s/f3wyo3hetpoisix/Screenshot%202016-04-13%2011.19.11.png?dl=0

Here is the steps how I setup this new project:
1. put the pod.spec file in project folder
2. pod init and modify the podfile like what you mentioned above and enabled 'use_framework'
3. install all the pods

**Comment by Raikerian** [comment]
Try to remove that line if you have Bolts installed with Pods.

**Comment by yaodong-li** [comment]
Already tried.

<img width="995" alt="screenshot 2016-04-13 21 25 29" src="https://cloud.githubusercontent.com/assets/6023208/14494624/4f5161d8-01be-11e6-88b3-77d2568d0cca.png">

**Comment by yaodong-li** [comment]
My steps are:
1.start a new project and put the pod.spec file in project folder
2. pod init and modify the podfile like what you mentioned above and enabled 'use_framework'
3. install all the pods

no manual installation of "qmservices" this time. Is that correct?

**Comment by Raikerian** [comment]
Try this one with Bolts as dependency (remove pod bolts from your podfile).

```
#
#  Be sure to run `pod spec lint QMServices.podspec --verbose --use-libraries --allow-warnings' to ensure this is a
#  valid spec and to remove all comments including this before submitting the spec.
#
#  To learn more about Podspec attributes see http://docs.cocoapods.org/specification.html
#  To see working Podspecs in the CocoaPods repo see https://github.com/CocoaPods/Specs/
#   
#  To submit use `pod trunk push QMServices.podspec --verbose --use-libraries --allow-warnings`
#


Pod::Spec.new do |s|
  s.name         = "QMServices"
  s.version      = "0.3.8"
  s.summary      = "Easy-to-use services for Quickblox SDK, for speeding up development of iOS chat applications."
  s.homepage     = "https://github.com/QuickBlox/q-municate-services-ios"
  s.license      = { :type => 'BSD', :file => '../LICENSE.txt' }
  s.authors      = {"Anton Sokolchenko" => "anton.sokolchenko@quickblox.com", "Andrey Ivanov" => "andrey.ivanov@quickblox.com", "Vitaliy Gorbachov" => "vitaliy.gorbachov@quickblox.com"}
  s.platform     = :ios, "7.0"
  s.source       = { :git => "https://github.com/QuickBlox/q-municate-services-ios.git", :tag => "#{s.version}"}
  s.source_files  = "Classes", "**/*.{h,m}"
  s.exclude_files = "Pods"
  s.frameworks = "CoreData"
  s.requires_arc = true
  s.xcconfig = { "FRAMEWORK_SEARCH_PATHS" => "$(PODS_ROOT)/QuickBlox/" }
  s.prefix_header_contents = 
  '#import <Quickblox/Quickblox.h>
#import <CoreData/CoreData.h>
#import <Quickblox/QBMulticastDelegate.h>
#import <Bolts/Bolts.h>'
  s.resource_bundle = {'QMChatCacheModel' => 'QMChatCache/QMChatCache/CoreData/QMChatServiceModel.xcdatamodeld', 'QMContactListCacheModel' => 'QMContactListCache/QMContactListCache/CoreData/QMContactListModel.xcdatamodeld', 'QMUsersCacheModel' => 'QMUsersCache/QMUsersCache/CoreData/QMUsersModel.xcdatamodeld'}
  s.dependency "Bolts",  '>= 1.5.0'
end
```

**Comment by yaodong-li** [comment]
Unfortunately same error

**Comment by Raikerian** [comment]
Have you tried to install bolts manually?

**Comment by yaodong-li** [comment]
Tried. No luck. 
At the same time, I'm using quickblox framework and write logics of caching etc instead

**Comment by Raikerian** [comment]
We have plans to push release on Friday (tomorrow). Fix should be available at that time.

**Comment by yaodong-li** [comment]
Thanks a lot. I would definitely check about that. Self-implementing a servicemanager in swift is very time-consuming

**Comment by Raikerian** [comment]
@fans3210 fixed in 0.3.9.

**Comment by yaodong-li** [comment]
Hi @Raikerian ,

Now issues come again in xcode 8 beta with swift 3. Attached are the two sample projects. 

 .  The first one is what I made before and it worked well in xcode 7 with swift 2.2. But crash in xcode 8 beta with swift 3.

https://www.dropbox.com/s/js0jeacs2f860jz/JsqumunicateTest_22Jun.zip?dl=0

 .  The second one is what I just created by following the "manual installation of QMServices". I also faced crash issue in this newly created project.

https://www.dropbox.com/s/gen91w3x8c0nbp4/ConnectedLife_22Jun.zip?dl=0

Is there any suggestion? I have to use manual installation because I have to use 'QMServices' as framework. Thanks a lot!
